File: app/src/main/java/com/najib/cryptowatcher/MainActivity.java
package com.najib.cryptowatcher;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import android.content.Intent;
import android.os.Bundle;


public class MainActivity extends AppCompatActivity implements SwipeRefreshLayout.OnRefreshListener {

    private CryptoAdapter mAdapter;
    private CryptoViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        viewModel = new ViewModelProvider(this, new ViewModelProvider.NewInstanceFactory()).get(CryptoViewModel.class);
        RecyclerView recyclerView = findViewById(R.id.recylerview);
        SwipeRefreshLayout swipeRefreshLayout  = findViewById(R.id.swipeRefresh);
        swipeRefreshLayout .setOnRefreshListener(this);
        mAdapter = new CryptoAdapter();
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(mAdapter);
        viewModel.requestData();

        viewModel.getData().observe(this, items ->{
            if (!items.isEmpty()){
                mAdapter.init(items, getApplicationContext());
                swipeRefreshLayout.setRefreshing(false);
            }
        });
        mAdapter.setOnItemClick(crypto -> startActivity(new Intent(this, DetailCryptoActivity.class).putExtra(DetailCryptoActivity.EXTRA_CRYPTO, crypto)));
    }

    @Override
    public void onRefresh() {
        viewModel.requestData();
    }
}

File: app/src/main/java/com/najib/cryptowatcher/MainActivity.kt
package com.najib.cryptowatcher

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProvider.NewInstanceFactory
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout.OnRefreshListener
import com.najib.cryptowatcher.databinding.ActivityMainBinding
import java.util.*

class MainActivity : AppCompatActivity(), OnRefreshListener {

    private lateinit var mAdapter: CryptoAdapter
    private lateinit var viewModel: CryptoViewModel
    private lateinit var binding: ActivityMainBinding


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        viewModel = ViewModelProvider(this, NewInstanceFactory()).get(CryptoViewModel::class.java)
        binding.swipeRefresh.setOnRefreshListener(this)
        mAdapter = CryptoAdapter()
        binding.recylerview.layoutManager = LinearLayoutManager(this)
        binding.recylerview.adapter = mAdapter
        viewModel.requestData()
        viewModel.data.observe(this, { items  ->
            if (items.isNotEmpty()) {
                mAdapter.init(items, applicationContext)
                binding.swipeRefresh.isRefreshing = false
            }
        })

        mAdapter.setOnItemClick(object : CryptoAdapter.OnItemClick{
            override fun click(crypto: Crypto?) {
                startActivity(
                    Intent(
                        this@MainActivity,
                        DetailCryptoActivity::class.java
                    ).putExtra(DetailCryptoActivity.EXTRA_CRYPTO, crypto)
                )
            }

        })

    }

    override fun onRefresh() {
        viewModel.requestData()
    }
}